instant-xml-macros/src/case.rs

```diff
@@ -0,0 +1,204 @@
+//! Originally from https://raw.githubusercontent.com/serde-rs/serde/master/serde_derive/src/internals/case.rs
+//! Code to convert the Rust-styled field/variant (e.g. `my_field`, `MyType`) to the
+//! case of the source (e.g. `my-field`, `MY_FIELD`).
+
+// See https://users.rust-lang.org/t/psa-dealing-with-warning-unused-import-std-ascii-asciiext-in-today-s-nightly/13726
+#[allow(deprecated, unused_imports)]
+use std::ascii::AsciiExt;
+
+use std::fmt::{self, Debug, Display};
+
+use self::RenameRule::*;
+
+/// The different possible ways to change case of fields in a struct, or variants in an enum.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum RenameRule {
+    /// Don't apply a default rename rule.
+    None,
+    /// Rename direct children to "lowercase" style.
+    LowerCase,
+    /// Rename direct children to "UPPERCASE" style.
+    UpperCase,
+    /// Rename direct children to "PascalCase" style, as typically used for
+    /// enum variants.
+    PascalCase,
+    /// Rename direct children to "camelCase" style.
+    CamelCase,
+    /// Rename direct children to "snake_case" style, as commonly used for
+    /// fields.
+    SnakeCase,
+    /// Rename direct children to "SCREAMING_SNAKE_CASE" style, as commonly
+    /// used for constants.
+    ScreamingSnakeCase,
+    /// Rename direct children to "kebab-case" style.
+    KebabCase,
+    /// Rename direct children to "SCREAMING-KEBAB-CASE" style.
+    ScreamingKebabCase,
+}
+
+impl Default for RenameRule {
+    fn default() -> Self {
+        None
+    }
+}
+
+static RENAME_RULES: &[(&str, RenameRule)] = &[
+    ("\"lowercase\"", LowerCase),
+    ("\"UPPERCASE\"", UpperCase),
+    ("\"PascalCase\"", PascalCase),
+    ("\"camelCase\"", CamelCase),
+    ("\"snake_case\"", SnakeCase),
+    ("\"SCREAMING_SNAKE_CASE\"", ScreamingSnakeCase),
+    ("\"kebab-case\"", KebabCase),
+    ("\"SCREAMING-KEBAB-CASE\"", ScreamingKebabCase),
+];
+
+impl RenameRule {
+    pub fn from_str(rename_all_str: &str) -> Result<Self, ParseError> {
+        for (name, rule) in RENAME_RULES {
+            if rename_all_str == *name {
+                return Ok(*rule);
+            }
+        }
+        Err(ParseError {
+            unknown: rename_all_str,
+        })
+    }
+
+    /// Apply a renaming rule to an enum variant, returning the version expected in the source.
+    pub fn apply_to_variant(&self, variant: &str) -> String {
```

Comment:
I think `apply_to_variant()` and `apply_to_field()` should take a `&Ident` so that we can yield the `ident.to_string()` allocation in case no change is needed.